File: fireplace-app/src/main/java/io/github/bric3/fireplace/ui/TitleBar.java
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.ui;

import com.github.weisj.darklaf.platform.SystemInfo;
import com.github.weisj.darklaf.platform.decorations.ExternalLafDecorator;

import javax.swing.*;
import java.awt.*;

public class TitleBar extends JPanel {
    public TitleBar(JComponent component) {
        setLayout(new BorderLayout());
        add(component, BorderLayout.CENTER);
    }

    @Override
    public void doLayout() {
        if (SystemInfo.isMac) {
            add(WindowButtonSpace.INSTANCE, BorderLayout.WEST);
        }
        if (SystemInfo.isWindows || SystemInfo.isLinux) {
            add(WindowButtonSpace.INSTANCE);
        }
        super.doLayout();
    }

    private static class WindowButtonSpace extends JComponent {
        private static final long serialVersionUID = 1L;
        private static final WindowButtonSpace INSTANCE = new WindowButtonSpace();

        private WindowButtonSpace() {}

        private Rectangle _windowButtonRect = null;

        private Rectangle windowButtonRect() {
            if (_windowButtonRect == null) {
                _windowButtonRect = ExternalLafDecorator.instance()
                                                        .decorationsManager()
                                                        .titlePaneLayoutInfo(getRootPane())
                                                        .windowButtonRect();
            }
            return _windowButtonRect;
        }

        public Dimension getPreferredSize() {
            var rectangle = windowButtonRect();
            var size = rectangle.getSize();
            if (SystemInfo.isMac) {
                size.width = size.width + rectangle.x;
            }
            if (SystemInfo.isWindows || SystemInfo.isLinux) {
                var rightAdjustment = getRootPane().getWidth() - windowButtonRect().x - windowButtonRect().width;
                size.width = size.width + rightAdjustment;
            }
            return size;
        }

        public Dimension getMinimumSize() {
            return getPreferredSize();
        }

        public Dimension getMaximumSize() {
            return getPreferredSize();
        }
    }
}

File: fireplace-app/src/main/kotlin/io/github/bric3/fireplace/ui/TitleBar.kt
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.ui

import com.github.weisj.darklaf.platform.SystemInfo
import com.github.weisj.darklaf.platform.decorations.ExternalLafDecorator
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.Rectangle
import javax.swing.JComponent
import javax.swing.JPanel

class TitleBar(component: JComponent) : JPanel(BorderLayout()) {
    init {
        add(component, BorderLayout.CENTER)
    }

    override fun doLayout() {
        if (SystemInfo.isMac) {
            add(WindowButtonSpace.INSTANCE, BorderLayout.WEST)
        }
        if (SystemInfo.isWindows || SystemInfo.isLinux) {
            add(WindowButtonSpace.INSTANCE)
        }
        super.doLayout()
    }

    private class WindowButtonSpace private constructor() : JComponent() {
        private val windowButtonRect: Rectangle by lazy {
            ExternalLafDecorator.instance()
                .decorationsManager()
                .titlePaneLayoutInfo(rootPane)
                .windowButtonRect()
        }

        override fun getPreferredSize(): Dimension {
            val size = windowButtonRect.size
            if (SystemInfo.isMac) {
                size.width = size.width + windowButtonRect.x
            }
            if (SystemInfo.isWindows || SystemInfo.isLinux) {
                val rightAdjustment = rootPane.width - windowButtonRect.x - windowButtonRect.width
                size.width = size.width + rightAdjustment
            }
            return size
        }

        override fun getMinimumSize(): Dimension {
            return preferredSize
        }

        override fun getMaximumSize(): Dimension {
            return preferredSize
        }

        companion object {
            private const val serialVersionUID = 1L
            val INSTANCE = WindowButtonSpace()
        }
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
